### ISSUE TITLE
Bug: [no-useless-default-assignment] False positives

### ISSUE THREAD

**Issue by arperry** [typescript-eslint/typescript-eslint#11846]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=JYWwDg9gTgLgBAbzgVwM4FMDCBDANrgI2wGMBrOAXzgDMoIQ4ByKdEmRgbgFgAoX4iADtU8ANrYANHAJTicALxNGAXQVNU9dCKiMAdKjC5gMABR7GASm58eA4fAD6IAJ458RMmrRY8hEqRNeODgTADc8ZHQALjhBZBACdChRVUUUqQgwJOwYaAB%2BGO1gQQBzCwUAPjhRXVrw3EiMrKgc6GUJIOrJaVllXiteXhhnLLgABWwWkFQ1MboQYAwAHiQAD0KYKGKSlMoK634hETgHMDpiLVQJqZnFbFRnQTkTMEnsaZjr99Ry%2BSqETp2Y5rGKiBzdBwyE5yRSMACMACYAMwqShqbAAd2wxjgrxu1go1iAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGYkACAXAngA4oDG0AlobgLQrzkB2uA9AwPbWzKJLLLUATRADMAhghqi%2B5AOYMAtoibooiaNDbRIAXxDagA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
import { useCallback } from 'react';

const [a, b, c = ''] = 'somestr'.split('.');

const _myCallback = useCallback(
  (value: number[] = [], operator?: string) => [...value, operator],
  [a, b, c]
);

type Params = Promise<{ x: string[] }>;

const _processParams = async (params: Params) => {
  const { x: [_a, _b, _c = '123'] } = await params;
};
```

### ESLint Config

```javascript
module.exports = {
  "rules": {"@typescript-eslint/no-useless-default-assignment": "error"}
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

None of these uses should get flagged under this rule.

### Actual Result

"Default value is useless because the (parameter|property) is not optional."

## Arrays are assumed to be of indefinite length

`const [a, b, c = ''] = 'somestr'.split('.');`
`const { x: [_a, _b, _c = '123'] } = await params;`

These are only useless assignments when the array is of sufficient length.

## Something about useCallback makes arg defaults get flagged

`(value: number[] = [], operator?: string) => [...value, operator]`

This works in isolation, but inside of useCallback, it gets flagged.

**Comment by kirkwaiblinger** [comment]
cc @ulrichstark

**Comment by kirkwaiblinger** [comment]
@arperry Thanks for the report!

> Arrays are assumed to be of indefinite length

This notably goes away with [`noUncheckedIndexedAccess`](https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess), and I see there was some intentional discussion around this on the PR: https://github.com/typescript-eslint/typescript-eslint/pull/11720#discussion_r2575894884. @ulrichstark or @ronami was this intentional as is?

> Something about useCallback makes arg defaults get flagged

Maybe a bug with generics? Probably needs some investigation... opened for PRs.

**Comment by ulrichstark** [comment]
Thanks for reporting! This is a slightly shorter repro:

```ts
// Case 1
const [a, b = ''] = 'somestr'.split('.');

// Case 2
declare function useCallback<T>(callback: T);
useCallback(
  (value: number[] = []) => { },
);

// Case 3
declare const params: string[];
const [c = '123'] = params;
```

[Playground](https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=PTAEGEEMGcFNQIwFgBQBjA9gO2gF1ANqQA0oARqALygDkNAulbdBgLax4BONAdNAA4AbAJa4AFLxoBKANypUICDHgAmVABNYaQZE7wAZgFcsaXMOyhDcKIMFlIaANYAeACoA%2BMWki37TgFygrrKoVrA2dg6OYqigoGIAbj6GsIFYhqxksJwEjNS5UlTuoADeoAC%2BxKghKApgUHCgAMwaWjp6oJg4%2BPy6kKzQgVzCWADmuXLo2HiEaEw0CCpNDEy9nP3QMkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tmUSWWS0AJogBmAQwR1xAygHMmAW0Qt0URNGgdokcGAC%2BIfUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)

**Comment by ulrichstark** [comment]
@kirkwaiblinger 

> This notably goes away with [noUncheckedIndexedAccess](https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess), and I see there was some intentional discussion around this on the PR: https://github.com/typescript-eslint/typescript-eslint/pull/11720#discussion_r2575894884.

I think we should not report these cases even if `noUncheckedIndexedAccess` is disabled.

**Comment by silverwind** [comment]
[Here](https://typescript-eslint.io/play/#ts=5.9.3&fileType=.tsx&code=GYVwdgxgLglg9mABAdwE4EMAOAeAKgPgApgwAuRXASnN0QG8AoRRVAUyhFSRIG4GBfBhAQBnKCgyZMrACaIAvBKyFCYEAFtya9QCNWqBYgCMlBfnpMW7Tkm19%2BlHkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tmUSWWS0AJogBmAQwR1xAygHMmAW0Qt0YANoAmALrgwAXxD6gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false) is another test case for a false-positive, I think it falls under "Case 2". I can only reproduce when using generics.

```ts
function wrap<T>(fn: T): T {
  return fn;
}
const wrapped = wrap((num: number = 1) => {
  return num;
});
```

**Comment by bradzacher** [comment]
How is "case 2" a false positive?
It looks like a correct report to me because the function parameter has a non-optional type.

Same applies to the case that silverwind showed - the parameter itself is non-optional so the default is unnecessary.

**Comment by arperry** [comment]
> How is "case 2" a false positive? It looks like a correct report to me because the function parameter has a non-optional type.
> 
> Same applies to the case that silverwind showed - the parameter itself is non-optional so the default is unnecessary.

Providing a default makes the argument optional.

**Comment by silverwind** [comment]
My intend is for `num` to be optional as signified by `= 1`. If I remove the wrapper, the rule no longer shows the error. I think the presence or absence of such wrappers should not have a impact on the lint result.

**Comment by silverwind** [comment]
Also if you check the type of `wrapped`, it is

- `(num?: number) => number` with `= 1`
- `(num: number) => number` without `= 1`

So the default is by definition not useless because it changes the resulting type.

**Comment by bradzacher** [comment]
Ah. Of course. It's been a minute since I've actually used default values in parameters so I'd forgotten that entirely. And you can't use `?:` with a default value as that is annoyingly a TS error.

**Comment by austinw-fineart** [comment]
Here's another false positive involving function overloads:
```ts
function foo(): void;
function foo(bar: null | string | undefined = undefined): void {
  console.log(bar);
}
```
Removing undefined causes `TS2394: This overload signature is not compatible with its implementation signature.`

**Comment by alumni** [comment]
Not sure if this is already covered by the deconstructing example above, but: 

```ts
async function findPosts({
    category,
    maxResults = 100, // not an error
}: {
    category: string;
    maxResults?: number;
}): Promise<PostEntity[]>;
```

Forgot to mention I have `strictNullChecks: false`, but this shouldn't affect anything since the property is optional.
